addon/serializer.js

```diff
@@ -127,27 +130,43 @@ export default DS.JSONAPISerializer.extend({
   __extractRelationships: function(modelClass, resourceHash) {
     const relationships = {};
 
-    modelClass.eachRelationship((key) => {
-      const relHash = resourceHash[key];
-      if (!relHash) { return; }
-
+    modelClass.eachRelationship((relName, {kind, type, options}) => {
       let data;
-      if (Ember.typeOf(relHash) === 'array') {
-        data = relHash.map((item) => this.__createRelationship(item['id'], key));
+
+      if (options.async) {
+        let suffix = kind === 'hasMany' ? 'Ids' : 'Id';
+        let key = Ember.String.singularize(relName) + suffix;
+        data = this.__buildRelationships(type, resourceHash[key], (elem) => elem);
       } else {
-        data = this.__createRelationship(relHash['id'], key);
+        data = this.__buildRelationships(type, resourceHash[relName], (elem) => elem.id);
       }
 
-      relationships[this.keyForRelationship(key)] = { 'data': data };
+      if (Ember.isPresent(data)) {
+        relationships[this.keyForRelationship(relName)] = { 'data': data };
+      }
     });
 
     return relationships;
   },
 
-  __createRelationship: function(id, key) {
+  __buildRelationships: function(type, collection, extractIdFn) {
+    if (!collection) {
+      return;
+    }
+
+    if (Ember.typeOf(collection) !== 'array') {
```

Comment:
Coud we flip this to:
```javascript
  if (Ember.typeOf(collection) === 'array') {
    return collection.map((elem) => {
       return this.__buildRelationship(extractIdFn(elem), type);
     });    
  } else {
     return this.__buildRelationship(extractIdFn(collection), type);
}
```

Asking if it's not a collection, then immediately referring to it as a collection is a bit jarring at first glance.